src/Exceptionless.Core/Services/StackService.cs

```diff
@@ -0,0 +1,94 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Exceptionless.Core.Extensions;
+using Exceptionless.Core.Repositories;
+using Foundatio.Caching;
+using Foundatio.Utility;
+using Microsoft.Extensions.Logging;
+
+namespace Exceptionless.Core.Services {
+    public class StackService {
+        private readonly ILogger<UsageService> _logger;
+        private readonly IStackRepository _stackRepository;
+        private readonly ICacheClient _cache;
+
+        public StackService(IStackRepository stackRepository, ICacheClient cache, ILoggerFactory loggerFactory = null) {
+            _stackRepository = stackRepository;
+            _cache = cache;
+            _logger = loggerFactory.CreateLogger<UsageService>();
+        }
+
+        public async Task IncrementStackUsageAsync(string organizationId, string projectId, string stackId, DateTime minOccurrenceDateUtc, DateTime maxOccurrenceDateUtc, int count) {
+            if (String.IsNullOrEmpty(organizationId) || String.IsNullOrEmpty(projectId) || String.IsNullOrEmpty(stackId) || count == 0)
+                return;
+            var expireTimeout = TimeSpan.FromDays(1);
+            var tasks = new List<Task>(4);
+
+            string occurenceCountCacheKey = GetStackOccurrenceCountCacheKey(organizationId, projectId, stackId),
+                occurrenceMinDateCacheKey = GetStackOccurrenceMinDateCacheKey(organizationId, projectId, stackId),
+                occurrenceMaxDateCacheKey = GetStackOccurrenceMaxDateCacheKey(organizationId, projectId, stackId),
+                occurrenceSetCacheKey = GetStackOccurrenceSetCacheKey();
+
+            var cachedOccurrenceMinDateUtc = await _cache.GetAsync<DateTime>(occurrenceMinDateCacheKey).AnyContext();
+            if (!cachedOccurrenceMinDateUtc.HasValue || cachedOccurrenceMinDateUtc.IsNull || cachedOccurrenceMinDateUtc.Value > minOccurrenceDateUtc)
+                tasks.Add(_cache.SetAsync(occurrenceMinDateCacheKey, minOccurrenceDateUtc, expireTimeout));
+
+            var cachedOccurrenceMaxDateUtc = await _cache.GetAsync<DateTime>(occurrenceMaxDateCacheKey).AnyContext();
+            if (!cachedOccurrenceMaxDateUtc.HasValue || cachedOccurrenceMinDateUtc.IsNull || cachedOccurrenceMaxDateUtc.Value < maxOccurrenceDateUtc)
+                tasks.Add(_cache.SetAsync(occurrenceMaxDateCacheKey, maxOccurrenceDateUtc, expireTimeout));
+
+            tasks.Add(_cache.IncrementAsync(occurenceCountCacheKey, count, expireTimeout));
+            tasks.Add(_cache.SetAddAsync(occurrenceSetCacheKey, Tuple.Create(organizationId, projectId, stackId), expireTimeout));
+
+            await Task.WhenAll(tasks).AnyContext();
+        }
+
+        public async Task SaveStackUsagesAsync(bool sendNotifications = true, CancellationToken cancellationToken = default) {
+            var occurrenceSetCacheKey = GetStackOccurrenceSetCacheKey();
+            var stackUsageSet = await _cache.GetSetAsync<Tuple<string, string, string>>(occurrenceSetCacheKey).AnyContext();
```

Comment:
I'm curious, do you know how Tuples round trips via binary serialization?

Reply:
The Tuples will be serialized same as POCO

Reply:
It looks like we are only adding to the set and never clearing it out which means it will have unbounded growth and this call will end up being a massive amount of data.

Reply:
I am thinking that maybe we can get the full set. Get the current count of the stack. Remove the item from the set. And then decrement occurrence count by the value that we had. Then even if at the same time the stack is being added back to the set it will be fine because it will be incrementing the count and we will be decrementing it. This approach requires us to add a SetRemove operation to the cache client.